test/studies/shootout/reverse-complement/bradc/revcomp-blc-gcc8.chpl

```diff
@@ -0,0 +1,170 @@
+/* The Computer Language Benchmarks Game
+   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
+
+   contributed by Brad Chamberlain
+   based on the C gcc #8 version by Jeremy Zerfas
+*/
+
+use IO, CTypes;
+
+config param readSize = 65536,
+             linesPerChunk = 8192;
+
+param eol = '\n'.toByte(),  // end-of-line, as an integer
+      cols = 61,            // # of characters per full row (including '\n')
+
+      // A 'bytes' value that stores the complement of each base at its index
+      cmpl = b"          \n                                                  "
+           + b"    TVGH  CD  M KN   YSAABW R       TVGH  CD  M KN   YSAABW R",
+             //    ↑↑↑↑  ↑↑  ↑ ↑↑   ↑↑↑↑↑↑ ↑       ↑↑↑↑  ↑↑  ↑ ↑↑   ↑↑↑↑↑↑ ↑
+             //    ABCDEFGHIJKLMNOPQRSTUVWXYZ      abcdefghijklmnopqrstuvwxyz
+
+      maxChars = cmpl.size; // upper bound on number of nucleotides used
+
+// map from pairs of nucleotide characters to their reversed complements
+var pairCmpl: [0..<join(maxChars, maxChars)] uint(16);
+
+// channels for doing efficient console I/O
+var stdinBin  = openfd(0).reader(iokind.native, locking=false,
+                           hints=ioHintSet.fromFlag(QIO_CH_ALWAYS_UNBUFFERED)),
+    stdoutBin = openfd(1).writer(iokind.native, locking=false,
+                           hints=ioHintSet.fromFlag(QIO_CH_ALWAYS_UNBUFFERED));
+
+proc main(args: [] string) {
+  // set up the 'pairCmpl' map
+  const chars = eol..<maxChars;
+  forall (i,j) in {chars, chars} do
+    pairCmpl[join(i,j)] = join(cmpl(j), cmpl(i));
+
+  // variables for reading into a dynamically growing buffer
+  var buffCap = readSize,       // capacity of our input buffer
+      buffDom = {0..<buffCap},  // index set for the input buffer
+      buff: [buffDom] uint(8),  // the buffer itself
+      readPos = 0;              // the current read position
+
+  do {
+    // read 'readSize' new characters
+    var newChars = stdinBin.readBinary(c_ptrTo(buff[readPos]), readSize),
+        nextSeqStart: int;
+
+    // if the new characters contain the start of the next sequence,
+    while findSeqStart(buff, readPos, newChars, nextSeqStart) {
+      // process this one
+      revcomp(buff, nextSeqStart);
+
+      // then shift the next sequence to the start of the buffer
+      newChars -= nextSeqStart - readPos + 1;
+
+      serial (nextSeqStart < newChars) do
+        forall j in 0..newChars do
+          buff[j] = buff[j+nextSeqStart];
+
+      // and reset to see whether there's another sequence ahead
+      readPos = 1;
```

Comment:
Should this reset to ``0``? Or maybe something like ``buff.domain.first``?

Reply:
I've bashed my head against this several times and been unable to come up with a version as clean as this one that passes all my corner case tests, though I feel as though an even cleaner version ought to exist.  The rationale for setting it to 1 here is that we've just shifted `>` to the 0th position, so there's no need to check that position again.  Probably setting it to 0 could also be made to work (since we have to check for `low==0` / `i != 0` within `findSeqStart()` currently anyway), but it would also require doing some +/- 1 logic on `newChars`.  This felt like a six / half-dozen case to me, and this felt like the cleanest working version I found.

I feel as though there ought to be a way to set up the very first call to `findSeqStart()` to remove the need for checking against `low` / `i` within the routine (by essentially setting it to the equivalent of this `1` here).  But so far I haven't been able to find an approach that works and is clean.

I'd be very happy if you/someone were able to crack this, though... I've just been unable to (it was the last thing I tried last night before giving up yet again).

Reply:
I wrestled with this more tonight and got a version that improves on things I think, effectively by moving the check for 0 into the callsite for findSeqStart().